README.md
# FEUP-8 - Open Source 8-bit Game Development Platform

## Motivation

[PICO-8](https://www.lexaloffle.com/pico-8.php) is a fantasy console for making, sharing and playing tiny games and other computer programs. When you turn it on, the machine greets you with a shell for typing in Lua commands and provides simple built-in tools for creating your own cartridges.

This tool was used during 2018 to teach young programmers with ages ranging from 11 to 14 years old at University of Porto's Junior University, and for dynamizing game construction at Game Jams. However, PICO-8 is a paid platform; and for that reason during the [Software Development Laboratory](https://sigarra.up.pt/feup/en/ucurr_geral.ficha_uc_view?pv_ocorrencia_id=281226) course we were tasked with presenting an open-source alternative, that also allows for better flexibility by being more suitable for education.

As such, we decided to implement a free open-source clone of PICO-8 adapted to our requirements. After conduction a [market analysis](https://gitlab.com/ldso18-19/t5g2/wikis/Market-analysis) we stumbled upon an open-source clone of PICO-8 already implemented, [TIC-80](https://github.com/nesbox/TIC-80/wiki), which we then decided to adapt and extend to our purposes, thus creating the FEUP-8 platform.

## Product vision

We expect to achieve a product that can be used in education of younger people as introduction to programming. The project will hopefully be used in the next iterations of FEUP's Junior University, replacing the aforementioned PICO-8, however with our work we believe that it can be used for more purposes as long as connected to education in the first contacts with programming.
Younger people will be attracted by the game itself and at the same time will be learning how to program in an introductory language like Lua.
Our product will have the advantage of being free and open source which allows everyone to use it and implement new features allowing it to improve over time.
In a most deep way our product will provide features beyond the TIC-80 educational resources, it will have a tests platform that allow teachers to make exercises and tests to it via a Web app. This way, younger developers can test his code to check if it is correct and receive feedback for it.

## Feedback

* Request a new feature on GitLab
* File a bug in [GitLab Issues](https://gitlab.com/ldso18-19/t5g2/issues).

## Development Team

This is being developed by the following team, feel free to contact any of us at any time.

| <img src="https://assets.gitlab-static.net/uploads/-/system/user/avatar/2846743/avatar.png" height="75px" width="75px"/>  | [`@afonsojramos`](https://github.com/afonsojramos)|        
<img src="https://assets.gitlab-static.net/uploads/-/system/user/avatar/2846741/avatar.png" height="75px" width="75px"/> | [`@NadiaCarvalho`](https://gitlab.com/NadiaCarvalho) | <img src="https://secure.gravatar.com/avatar/40f5ded910f5f5a744db1ad56887a46a?s=180&d=identicon" height="75px" width="75px"/>  | [`@up201503995`](https://gitlab.com/up201503995)| <img src="https://secure.gravatar.com/avatar/147062fb337b447edbe58d1a6c16f458?s=180&d=identicon"      height="75px" width="75px"/> | [`@tiagolascasas`](https://gitlab.com/tiagolascasas)
--|--|--|--|--|--|--|--|
| <img src="https://secure.gravatar.com/avatar/f492e5707cbf0fca0f3225e0fcdbf52b?s=180&d=identicon" height="75px" width="75px"/>  | [`@Lantaros`](https://gitlab.com/Lantaros)| <img src="https://secure.gravatar.com/avatar/8dacc894429b1c487a94acfbe447c078?s=180&d=identicon"      height="75px" width="75px"/> | [`@TrocaTudo95`](https://gitlab.com/TrocaTudo95) | <img src="https://secure.gravatar.com/avatar/b9f320d003766e6ab9e859c6364d7744?s=180&d=identicon" height="75px" width="75px"/>  | [`@up201504196`](https://gitlab.com/up201504196)| <img src="http://4sqday.com/wp-content/uploads/2012/07/Pointing_Finger.jpg"   height="75px" width="75px"/> | YOU

## Build instructions

Since there are two main components to our project, we present their building instructions separately.

### FEUP-8 fantasy console

Regardless of the OS, make sure that you are on the ```FEUP-8``` subfolder of the project during **every stage** of the following instructions.

#### Linux (Ubuntu 18.04.1 LTS)

Firstly, install CMake:
```
wget "https://cmake.org/files/v3.12/cmake-3.12.0-Linux-x86_64.sh"
sudo sh cmake-3.12.0-Linux-x86_64.sh --skip-license --prefix=/usr
```

Then, run the following commands in the terminal
```
sudo apt-get install git cmake libgtk-3-dev libglvnd-dev:i386 libglvnd-dev libglu-dev lcov -y
git clone --recursive https://gitlab.com/ldso18-19/t5g2 && cd t5g2/FEUP-8
cmake . && make -j4
```

A binary executable will be generated in /bin.

After the initial install, simply run the ```build_and_run.sh``` script every time you want to compile and run.

These instructions, as said on the title, were only verified on Ubuntu 18.04.1 LTS. Other distros and versions may have slight differences in dependencies or package manager.

#### Mac

Firstly, install `Command Line Tools for Xcode` and `brew` package manager

Then, run the following commands in the Terminal
```
brew install git cmake
git clone --recursive https://gitlab.com/ldso18-19/t5g2
cd t5g2/FEUP-8
cmake . && make -j4
```

A binary executable will be generated in ```/bin```.

After the initial install, simply run the ```build_and_run_mac.sh``` script every time you want to compile and run.

#### Windows

Windows builds are no longer supported. The code coverage report generation required additional configuration of that build, and since Windows builds are not within the current scope of the project, we decided to deprecate it instead.

#### To run the tests and generate coverage report:

```
sh build_and_test.sh    # to run tests
sh run_coverage.sh      # to generate a coverage report
```

A coverage report will be generated in ```/tic80_coverage```.

#### To run the linter/automatic formatter:

```
sh lint.sh
```

If you don't have ```clang-format``` installed, the script will tell you in a friendly manner that you need to get it first.

On Ubuntu, this can be installed using ```sudo apt-get install clang-format```, though the package name may change in other distros or OS.

### FEUP-8 Website (Linux and Mac only)

Regardless of the OS, make sure that you are on the ```Website``` subfolder of the project during **every stage** of the following instructions.

#### To run the website locally:

Install the following dependencies using your distro's package manager. The following list has been validated using apt-get on Ubuntu 18.04.1 LTS, and it may differ slightly on other distros or OS. Install them one by one on this exact order to ensure maximal compatibility:

- php7.2
- php-xdebug
- composer
- php7.2-sqlite
- php7.2-mbstring
- php7.2-xml
- lua50
- luajit
- luarocks

Then, run the initial build:

```
git clone --recursive https://gitlab.com/ldso18-19/t5g2.git
cd t5g2/Website
sudo composer install   # use sudo only if you get permission errors, and only after you verify if you can't solve them with chmod
luarocks install luaunit
```

Now, you need to change the ```DB_DATABASE``` path in the ```.env``` file to the one on your local machine. Once that is done, run the following:

```
sudo sh setup.sh    # use sudo only if you get permission errors, and only after you verify if you can't solve them with chmod
php artisan serve
```

After the former first time setup, simply run:

```
php artisan serve
```

#### To run the unit tests and generate the coverage report:

```
sudo sh setup_tests.sh
vendor/bin/phpunit
```

A code coverage report will be generated on ```/report```.

#### To run the linter and automatic formatter:

```
sh lint.sh
```

## Web API Documentation 

The documentation  of the API that allows communication between the FEUP-8 app and the server can be found here:

**https://app.swaggerhub.com/apis-docs/dinisvale/LDSO/1.0.0**


## Related Projects

Please refer to our Wiki page on [market analysis](https://gitlab.com/ldso18-19/t5g2/wikis/Market-analysis) to know more about other projects similar to FEUP-8.


## License

Licensed under the [MIT](LICENSE) License.


changelog.md
# Changelog
All notable changes to this project will be documented in this file.


## [0.1.0] - 15-10-2018 - Sprint 1
### Closed PBI's
- Database Schema #34
- Database SQL #32
- Setup a Unit Test framework for FEUP-8 #39
- Deploy laravel framewok and structure the website #38
- Standardize an exercise format #41
- Database migrations #53
- Study of the TIC-80 App #37

### Increment description
We started by studying the TIC-80 app in order to understand the architecture so that we could integrate our code to expand the app and make it more suitable for education.
Before starting the development of the code, we got a framework for unit testing it.
On server side, we started by architecturing the app, setting up the framework to be used (laravel) and developing the database schema and sql code.
We started the development of the API responsible by the communication between the app and the server.
In the end of this sprint it was only possible to list the exercises in the database using the app. However, communication and server backend was mostly developed for getExericiseDetails although not finished and therefore not possible for the user to see happening.



## [0.2.0] - 30-10-2018 - Sprint 2
### Closed PBI's
- Execute lua scripts on server #51
- Update Docker image #63
- Run test on exercise as a Student #16
- FEUP-8 API #36
- List existent exercises as a Student #7
- Solve exercise as a Student #25
- Docker and Gitlab CI #46
- Create lua exercises #54
- Gitlab CI #57
- Docker Application FEUP-8 #61
- Create Gitlab templates #60
- Server tests parsing #52
- Application Unit Tests #49
- Prototype a "see exercise" interface in FEUP-8 #40
- Website API #35
- Open exercise as a Student #13

### Increment description
We started by setting up the Gitlab CI so we could have testing feedback on our code on each commit. We also set up the docker images for the client app and server. We enabled the server to be able to run lua code in order to run the tests on an exercise.
In the end of this sprint the user was able to list the details of an exercise on the client app. Was also implemented the backend to run tests on an exercise (which implied sending the code to the server, running the tests on it and sending the results back to the client) and save their progress on an exercise to the server so that they can start from the point they finished at, the next time they login in on the same account, on any device, allowing synchronization.



## [0.3.0] - 14-11-2018 - Sprint 3
### Closed PBI's
- Open-API Documentation #74
- Fix bugs on tests output #75
- Decide which FEUP-8 files should be part of the coverage #59
- Laravel unit tests #50
- Fix the problems left from previous sprint #73
- Add the xdebug install options to Dockerfile #69
- Add descriptive acceptance tests to sprint 3 PBIs #67
- Add linter to application (C) #55
- Add clang-format linter to C files #68 
- API unit tests #48
- Show test output result on interface #71
- Read test output as a student #18
- Unit tests on client API #64
- Website-API new functionalities #62
- Coverage tests for laravel #58

### Increment description
This sprint was more focused in fixes and new  related to the SCRUM process that is being followed. We have finished all unit tests to the code developed so far, as well as tested it for coverage.
The code is now also verified by a linter in each push.
The API developed in previous sprints is now fully documented using the Open-API swagger hub platform.
However, we have also finished the development of the functionalities that allows a student (user of the program) to test student code on server and show tests results.


## [0.4.0] - 26-11-2018 - Sprint 4
### Closed PBI's
- Added automated deployment #66
- List existent exercises as a Teacher #10
- Create exercise as a Teacher #9
- Add tips to unit tests as a Teacher #17
- Edit exercise as a Teacher #8
- Login as a Teacher on the Website #22
- List the exercise I've created #24

### Increment description
This sprint was dedicated to the website, allowing a teacher to add exercises that can later be used in the FEUP-8. A teacher can now, using the website, login, create an exercise and respective tests, as well as edit them. The teacher can also list all the exercises available in the platform, as well as his own exercises (i.e. the exercises developed by him).
The continuous deployment, started in the last sprint, was also finished.


## [1.0.0] - 12-12-2018 - Sprint 5
### Closed PBI's
- Add right permissions to Actions on Website #94
- Improve interoperability between the two components #95
- Login as a Student on FEUP-8 #11
- Register on Website as a Visitor #20
- Logout as a Student #12
- Register on the FEUP-8 application as a Visitor #21
- Change Database Management System #90
- Delete an exercise as a Teacher #26
- Open a saved exercise as a Student #15
- Website Tuning #23
- Save an exercise progress as a Student #14
- Convert TIC-80 into FEUP-8 #47

### Increment description
This sprint focused on finishing the implementation of important features such as support for saving and loading progress feature on client-side (on server side it was already implemented). Using the FEUP-8 app, a student can now login, logout and register.

On the Website, it is now possible to delete an exercise when logged in as a Teacher. The Website aesthetics and navbar were also improved.

As final fixes, we improved the client and server interoperability for final release:
* fixed FEUP-8 client not working with production server
* the FEUP-8 app is now more configurable as the server Domain/IP address can be changed by editing a config file (conf.ini)
* FEUP-8 error messages related to communication between the client and the server were improved

#### Bug Fixes:
	FEUP-8:
		* Segmentation fault on trying to load an exercise with id non positive
		* Segmentation fault on load tests from an exercise that has none

FEUP-8/src/menu.c
// MIT License

// Copyright (c) 2017 Vadim Grigoruk @nesbox // grigoruk@gmail.com

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "menu.h"
#include "fs.h"

#define DIALOG_WIDTH (TIC80_WIDTH/2)
#define DIALOG_HEIGHT (TIC80_HEIGHT/2)

static const char* Rows[] =
{
	"RESUME GAME",
	"RESET GAME",
	"GAMEPAD CONFIG",
	"",
	"CLOSE GAME",
	"QUIT FEUP-8",
};

static void resumeGame(Menu* menu)
{
	hideGameMenu();
}

static void resetGame(Menu* menu)
{
	tic_mem* tic = menu->tic;

	tic->api.reset(tic);

	setStudioMode(TIC_RUN_MODE);
}

static void gamepadConfig(Menu* menu)
{
	playSystemSfx(2);
	menu->mode = GAMEPAD_MENU_MODE;
	menu->gamepad.tab = 0;
}

static void closeGame(Menu* menu)
{
	exitFromGameMenu();
}

static void(*const MenuHandlers[])(Menu*) = {resumeGame, resetGame, gamepadConfig, NULL, closeGame, exitStudio};

static tic_rect getRect(Menu* menu)
{
	tic_rect rect = {(TIC80_WIDTH - DIALOG_WIDTH)/2, (TIC80_HEIGHT - DIALOG_HEIGHT)/2, DIALOG_WIDTH, DIALOG_HEIGHT};

	rect.x -= menu->pos.x;
	rect.y -= menu->pos.y;

	return rect;
}
static void drawDialog(Menu* menu)
{
	tic_rect rect = getRect(menu);

	tic_mem* tic = menu->tic;

	tic_rect header = {rect.x, rect.y-(TOOLBAR_SIZE-1), rect.w, TOOLBAR_SIZE};

	if(checkMousePos(&header))
	{
		setCursor(tic_cursor_hand);

		if(checkMouseDown(&header, tic_mouse_left))
		{
			if(!menu->drag.active)
			{
				menu->drag.start.x = getMouseX() + menu->pos.x;
				menu->drag.start.y = getMouseY() + menu->pos.y;

				menu->drag.active = true;
			}
		}
	}

	if(menu->drag.active)
	{
		setCursor(tic_cursor_hand);

		menu->pos.x = menu->drag.start.x - getMouseX();
		menu->pos.y = menu->drag.start.y - getMouseY();

		tic_rect rect = {0, 0, TIC80_WIDTH, TIC80_HEIGHT};
		if(!checkMouseDown(&rect, tic_mouse_left))
			menu->drag.active = false;
	}

	rect = getRect(menu);

	tic->api.rect(tic, rect.x, rect.y, rect.w, rect.h, (tic_color_blue));
	tic->api.rect_border(tic, rect.x, rect.y, rect.w, rect.h, (tic_color_white));
	tic->api.line(tic, rect.x, rect.y+DIALOG_HEIGHT, rect.x+DIALOG_WIDTH-1, rect.y+DIALOG_HEIGHT, (tic_color_black));
	tic->api.rect(tic, rect.x, rect.y-(TOOLBAR_SIZE-2), rect.w, TOOLBAR_SIZE-2, (tic_color_white));
	tic->api.line(tic, rect.x+1, rect.y-(TOOLBAR_SIZE-1), rect.x+DIALOG_WIDTH-2, rect.y-(TOOLBAR_SIZE-1), (tic_color_white));

	{
		static const char Label[] = "GAME MENU";
		s32 size = tic->api.text(tic, Label, 0, -TIC_FONT_HEIGHT, 0, false);
		tic->api.text(tic, Label, rect.x + (DIALOG_WIDTH - size)/2, rect.y-(TOOLBAR_SIZE-2), (tic_color_gray), false);
	}

	{
		u8 chromakey = 14;
		tic->api.sprite_ex(tic, &getConfig()->cart->bank0.tiles, 0, rect.x+6, rect.y-4, 2, 2, &chromakey, 1, 1, tic_no_flip, tic_no_rotate);
	}
}

static void drawTabDisabled(Menu* menu, s32 x, s32 y, s32 id)
{
	enum{Width = 15, Height = 7};
	tic_mem* tic = menu->tic;

	tic_rect rect = {x, y, Width, Height};
	bool over = false;

	if(menu->gamepad.tab != id && checkMousePos(&rect))
	{
		setCursor(tic_cursor_hand);
		over = true;

		if(checkMouseDown(&rect, tic_mouse_left))
		{
			menu->gamepad.tab = id;
			menu->gamepad.selected = -1;
			return;
		}
	}

	tic->api.rect(tic, x, y-1, Width, Height+1, (tic_color_dark_gray));
	tic->api.pixel(tic, x, y+Height-1, (tic_color_blue));
	tic->api.pixel(tic, x+Width-1, y+Height-1, (tic_color_blue));

	{
		char buf[] = "#1";
		sprintf(buf, "#%i", id+1);
		tic->api.fixed_text(tic, buf, x+2, y, (over ? tic_color_light_blue : tic_color_gray), false);
	}
}

static void drawTab(Menu* menu, s32 x, s32 y, s32 id)
{
	enum{Width = 15, Height = 7};
	tic_mem* tic = menu->tic;

	tic->api.rect(tic, x, y-2, Width, Height+2, (tic_color_white));
	tic->api.pixel(tic, x, y+Height-1, (tic_color_black));
	tic->api.pixel(tic, x+Width-1, y+Height-1, (tic_color_black));
	tic->api.rect(tic, x+1, y+Height, Width-2 , 1, (tic_color_black));

	{
		char buf[] = "#1";
		sprintf(buf, "#%i", id+1);
		tic->api.fixed_text(tic, buf, x+2, y, (tic_color_gray), false);
	}
}

static void drawPlayerButtons(Menu* menu, s32 x, s32 y)
{
	tic_mem* tic = menu->tic;

	u8 chromakey = 0;

	tic_key* codes = getKeymap();

	enum {Width = 41, Height = TIC_SPRITESIZE, Rows = 4, Cols = 2, MaxChars = 5, Buttons = 8};

	for(s32 i = 0; i < Buttons; i++)
	{
		tic_rect rect = {x + i / Rows * (Width+2), y + (i%Rows)*(Height+1), Width, TIC_SPRITESIZE};
		bool over = false;

		s32 index = i+menu->gamepad.tab * Buttons;

		if(checkMousePos(&rect))
		{
			setCursor(tic_cursor_hand);

			over = true;

			if(checkMouseClick(&rect, tic_mouse_left))
			{
				menu->gamepad.selected = menu->gamepad.selected != index ? index : -1;
			}
		}

		if(menu->gamepad.selected == index && menu->ticks % TIC_FRAMERATE < TIC_FRAMERATE / 2)
			continue;

		tic->api.sprite_ex(tic, &getConfig()->cart->bank0.tiles, 8+i, rect.x, rect.y, 1, 1, &chromakey, 1, 1, tic_no_flip, tic_no_rotate);

		static const char* const Names[tic_keys_count] = {"...", "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "minus", "equals", "leftbracket", "rightbracket", "backslash", "semicolon", "apostrophe", "grave", "comma", "period", "slash", "space", "tab", "return", "backspace", "delete", "insert", "pageup", "pagedown", "home", "end", "up", "down", "left", "right", "capslock", "ctrl", "shift", "alt", "escape", "f1", "f2", "f3", "f4", "f5", "f6", "f7", "f8", "f9", "f10", "f11", "f12"};

		s32 code = codes[index];
		char label[32];
		strcpy(label, Names[code]);

		if(strlen(label) > MaxChars)
			label[MaxChars] = '\0';

		tic->api.text(tic, label, rect.x+10, rect.y+2, (over ? tic_color_gray : tic_color_black), false);
	}
}

static void drawGamepadSetupTabs(Menu* menu, s32 x, s32 y)
{
	enum{Width = 90, Height = 41, Tabs = TIC_GAMEPADS, TabWidth = 16};
	tic_mem* tic = menu->tic;

	tic->api.rect(tic, x, y, Width, Height, (tic_color_white));
	tic->api.pixel(tic, x, y, (tic_color_blue));
	tic->api.pixel(tic, x+Width-1, y, (tic_color_blue));
	tic->api.pixel(tic, x, y+Height-1, (tic_color_black));
	tic->api.pixel(tic, x+Width-1, y+Height-1, (tic_color_black));
	tic->api.rect(tic, x+1, y+Height, Width-2 , 1, (tic_color_black));

	for(s32 i = 0; i < Tabs; i++)
		(menu->gamepad.tab == i ? drawTab : drawTabDisabled)(menu, x + 73 - i*TabWidth, y + 43, i);

	drawPlayerButtons(menu, x + 3, y + 3);
}

static void drawGamepadMenu(Menu* menu)
{
	drawDialog(menu);

	tic_mem* tic = menu->tic;

	tic_rect dlgRect = getRect(menu);

	static const char Label[] = "BACK";

	tic_rect rect = {dlgRect.x + 25, dlgRect.y + 56, (sizeof(Label)-1)*TIC_FONT_WIDTH, TIC_FONT_HEIGHT};

	bool over = false;
	bool down = false;

	if(checkMousePos(&rect))
	{
		setCursor(tic_cursor_hand);

		over = true;

		if(checkMouseDown(&rect, tic_mouse_left))
		{
			down = true;
		}

		if(checkMouseClick(&rect, tic_mouse_left))
		{
			menu->gamepad.selected = -1;
			menu->mode = MAIN_MENU_MODE;
			playSystemSfx(2);
			return;
		}
	}

	if(down)
	{
		tic->api.text(tic, Label, rect.x, rect.y+1, (tic_color_light_blue), false);
	}
	else
	{
		tic->api.text(tic, Label, rect.x, rect.y+1, (tic_color_black), false);
		tic->api.text(tic, Label, rect.x, rect.y, (over ? tic_color_light_blue : tic_color_white), false);
	}

	{
		static const u8 Icon[] =
		{
			0b10000000,
			0b11000000,
			0b11100000,
			0b11000000,
			0b10000000,
			0b00000000,
			0b00000000,
			0b00000000,
		};

		drawBitIcon(rect.x-7, rect.y+1, Icon, (tic_color_black));
		drawBitIcon(rect.x-7, rect.y, Icon, (tic_color_white));
	}

	drawGamepadSetupTabs(menu, dlgRect.x+25, dlgRect.y+4);
}

static void drawMainMenu(Menu* menu)
{
	tic_mem* tic = menu->tic;

	drawDialog(menu);

	tic_rect rect = getRect(menu);

	{
		for(s32 i = 0; i < COUNT_OF(Rows); i++)
		{
			if(!*Rows[i])continue;

			tic_rect label = {rect.x + 22, rect.y + (TIC_FONT_HEIGHT+1)*i + 16, 86, TIC_FONT_HEIGHT+1};
			bool over = false;
			bool down = false;

			if(checkMousePos(&label))
			{
				setCursor(tic_cursor_hand);

				over = true;

				if(checkMouseDown(&label, tic_mouse_left))
				{
					down = true;
					menu->main.focus = i;
				}

				if(checkMouseClick(&label, tic_mouse_left))
				{
					MenuHandlers[i](menu);
					return;
				}
			}

			if(down)
			{
				tic->api.text(tic, Rows[i], label.x, label.y+1, (tic_color_light_blue), false);
			}
			else
			{
				tic->api.text(tic, Rows[i], label.x, label.y+1, (tic_color_black), false);
				tic->api.text(tic, Rows[i], label.x, label.y, (over ? tic_color_light_blue : tic_color_white), false);
			}

			if(i == menu->main.focus)
			{
				static const u8 Icon[] =
				{
					0b10000000,
					0b11000000,
					0b11100000,
					0b11000000,
					0b10000000,
					0b00000000,
					0b00000000,
					0b00000000,
				};

				drawBitIcon(label.x-7, label.y+1, Icon, (tic_color_black));
				drawBitIcon(label.x-7, label.y, Icon, (tic_color_white));
			}
		}
	}
}

static void processGamedMenuGamepad(Menu* menu)
{
	if(menu->gamepad.selected < 0)
	{
		enum
		{
			Up, Down, Left, Right, A, B, X, Y
		};

		if(menu->tic->api.btnp(menu->tic, A, -1, -1))
		{
			menu->mode = MAIN_MENU_MODE;
			playSystemSfx(2);
		}
	}
}

static void processMainMenuGamepad(Menu* menu)
{
	enum{Count = COUNT_OF(Rows), Hold = 30, Period = 5};

	enum
	{
		Up, Down, Left, Right, A, B, X, Y
	};

	if(menu->tic->api.btnp(menu->tic, Up, Hold, Period))
	{
		do
		{
			if(--menu->main.focus < 0)
				menu->main.focus = Count - 1;

			menu->main.focus = menu->main.focus % Count;
		} while(!*Rows[menu->main.focus]);

		playSystemSfx(2);
	}

	if(menu->tic->api.btnp(menu->tic, Down, Hold, Period))
	{
		do
		{
			menu->main.focus = (menu->main.focus+1) % Count;
		} while(!*Rows[menu->main.focus]);

		playSystemSfx(2);
	}

	if(menu->tic->api.btnp(menu->tic, A, -1, -1))
	{
		MenuHandlers[menu->main.focus](menu);
	}
}

static void saveMapping(Menu* menu)
{
	fsSaveRootFile(menu->fs, KEYMAP_DAT_PATH, getKeymap(), KEYMAP_SIZE, true);
}

static void processKeyboard(Menu* menu)
{
	tic_mem* tic = menu->tic;

	if(tic->ram.input.keyboard.data == 0) return;

	if(menu->gamepad.selected < 0)
		return;

	if(keyWasPressed(tic_key_escape));
	else if(anyKeyWasPressed())
	{
		for(s32 i = 0; i < TIC80_KEY_BUFFER; i++)
		{
			tic_key key = tic->ram.input.keyboard.keys[i];

			if(tic->api.keyp(tic, key, -1, -1))
			{
				tic_key* codes = getKeymap();
				codes[menu->gamepad.selected] = key;
				saveMapping(menu);
				break;
			}
		}
	}

	menu->gamepad.selected = -1;
}

static void tick(Menu* menu)
{
	menu->ticks++;

	processKeyboard(menu);

	if(getStudioMode() != TIC_MENU_MODE)
		return;

	if(!menu->init)
	{
		playSystemSfx(0);

		menu->init = true;
	}

	memcpy(menu->tic->ram.vram.screen.data, menu->bg, sizeof menu->tic->ram.vram.screen.data);

	switch(menu->mode)
	{
	case MAIN_MENU_MODE:
		processMainMenuGamepad(menu);
		drawMainMenu(menu);
		break;
	case GAMEPAD_MENU_MODE:
		processGamedMenuGamepad(menu);
		drawGamepadMenu(menu);
		break;
	}
}

void initMenu(Menu* menu, tic_mem* tic, FileSystem* fs)
{
	*menu = (Menu)
	{
		.init = false,
		.fs = fs,
		.tic = tic,
		.tick = tick,
		.bg = menu->bg,
		.ticks = 0,
		.pos = {0, 0},
		.main =
		{
			.focus = 0,
		},
		.gamepad =
		{
			.tab = 0,
			.selected = -1,
		},
		.drag =
		{
			.start = {0, 0},
			.active = 0,
		},
		.mode = MAIN_MENU_MODE,
	};

	enum{Size = sizeof tic->ram.vram.screen.data};

	if(!menu->bg)
		menu->bg = malloc(Size);

	if(menu->bg)
		memcpy(menu->bg, tic->ram.vram.screen.data, Size);
}
